src/analysis/retail/rogue/assassination/modules/talents/CrimsonTempestUptimeAndSnapshots.tsx

```diff
@@ -0,0 +1,227 @@
+    // const snapshotNames = snapshots.map((ss) => ss.name);
+    // const prevSnapshotNames = prevSnapshots === null ? null : prevSnapshots.map((ss) => ss.name);
+    const wasUnacceptableDowngrade =
+      prevPower > power && remainingOnPrev > SNAPSHOT_DOWNGRADE_BUFFER;
+    const wasUpgrade = prevPower < power;
+
+    let snapshotPerformance: QualitativePerformance = QualitativePerformance.Good;
+    let snapshotSummary = <div>Good snapshot usage</div>;
+    let snapshotDetails = <div>Good snapshot usage.</div>;
+    if (wasUnacceptableDowngrade) {
+      snapshotPerformance = QualitativePerformance.Fail;
+      snapshotSummary = <div>Unacceptable downgrade of snapshot</div>;
+      snapshotDetails = (
+        <div>
+          Unacceptable downgrade of snapshot. Try not to overwrite your snapshotted Crimson Tempest
+          unless it's within the last {formatDurationMillisMinSec(SNAPSHOT_DOWNGRADE_BUFFER)}.
+        </div>
+      );
+    }
+    if (clipped > 0) {
+      snapshotPerformance = wasUpgrade ? QualitativePerformance.Ok : QualitativePerformance.Fail;
+      snapshotSummary = wasUpgrade ? (
+        <div>Clipped but upgraded existing snapshotted Crimson Tempest</div>
+      ) : (
+        <div>Clipped existing snapshotted Crimson Tempest</div>
+      );
+      snapshotDetails = wasUpgrade ? (
+        <div>
+          Clipped but upgraded existing snapshotted Crimson Tempest. Try not to clip your
+          snapshotted Crimson Tempest.
+        </div>
+      ) : (
+        <div>
+          Clipped existing snapshotted Crimson Tempest. Try not to clip your snapshotted Crimson
+          Tempest.
+        </div>
+      );
+    }
+
+    const checklistItems: ChecklistUsageInfo[] = [
+      {
+        check: 'snapshot',
+        timestamp: cast.timestamp,
+        performance: snapshotPerformance,
+        summary: snapshotSummary,
+        details: snapshotDetails,
+      },
+    ];
+
+    // const tooltip = (
+    //   <>
+    //     @ <strong>{this.owner.formatTimestamp(cast.timestamp)}</strong> targetting{' '}
+    //     <strong>{targetName || 'unknown'}</strong>
+    //     <br />
+    //     {prevSnapshotNames !== null && (
+    //       <>
+    //         Refreshed on target w/ {(remainingOnPrev / 1000).toFixed(1)}s remaining{' '}
+    //         {clipped > 0 && (
+    //           <>
+    //             <strong>- Clipped {(clipped / 1000).toFixed(1)}s!</strong>
+    //           </>
+    //         )}
+    //         <br />
+    //       </>
+    //     )}
+    //     Snapshots: <strong>{snapshotNames.length === 0 ? 'NONE' : snapshotNames.join(', ')}</strong>
+    //     <br />
+    //     {prevSnapshotNames !== null && (
+    //       <>
+    //         Prev Snapshots:{' '}
+    //         <strong>
+    //           {prevSnapshotNames.length === 0 ? 'NONE' : prevSnapshotNames.join(', ')}
+    //         </strong>
+    //       </>
+    //     )}
+    //   </>
+    // );
```

Comment:
deleto

also starting to thing that *maybe* there is some room for a little more shared code here? doesn't need to be part of this PR since it is already HUGE

Reply:
I'll see what all I can extract out here, but these snapshot bits are remnants from the old Assassination code that I need to embed in the spell usage details. I just need to find a way to make it look good.